### PR TITLE
Support multiple AmazonCloudWatchAgent CRs and add integration for Prometheus Target Allocator

### PR DESCRIPTION
**Pull request by okankoAMZ**
*Description of changes:*
* Adds support for configuring multiple AmazonCloudWatchAgent workloads, each of which can be a deployment/daemonset/statefulset
* Adds support for Prometheus monitoring via the CloudWatchAgent when using the helm-chart / EKS add-on
* Adds support for associating a AmazonCloudWatchAgent workload with a corresponding Prometheus TargetAllocator installation, which will be managed by the CloudWatchAgentOperator

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-custom-resource.yaml [hunk 1]
```diff
@@ -44,40 +47,66 @@ data:
...
-  {{- with .Values.agent.resources }}
-  resources: {{- toYaml . | nindent 4}}
+  {{- if $customAgent.prometheus.targetAllocator }}
+  targetAllocator:
+    enabled: {{ $customAgent.prometheus.targetAllocator.enabled | default false }}
+    image: {{ include "target-allocator.modify-image" (merge (dict "image" $customAgent.prometheus.targetAllocator.image) $ ) }}
+    allocationStrategy: "consistent-hashing"
+    {{- if $customAgent.prometheus.targetAllocator.prometheusCR }}
+    prometheusCR: {{ $customAgent.prometheus.targetAllocator.prometheusCR.enabled | default false }}
+    {{- end }}
```
Comment (musa-asad) [inline charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-custom-resource.yaml:100]:
```suggestion
    {{- end }}
    {{- else if $.Values.agent.prometheus.targetAllocator }}
  targetAllocator:
    enabled: {{ $.Values.agent.prometheus.targetAllocator.enabled | default false }}
    image: {{ include "target-allocator.modify-image" (merge (dict "image" $.Values.agent.prometheus.targetAllocator.image) $ ) }}
    allocationStrategy: "consistent-hashing"
    {{- if $.Values.agent.prometheus.targetAllocator.prometheusCR }}
    prometheusCR: {{ $.Values.agent.prometheus.targetAllocator.prometheusCR.enabled | default false }}
    {{- end }}
```

File: charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-custom-resource.yaml [hunk 2]
```diff
@@ -44,40 +47,66 @@ data:
...
   {{- end }}
-  {{- if .Values.agent.otelConfig }}
-  otelConfig: {{ include "cloudwatch-agent.modify-yaml-config" (merge (dict "OtelConfig" .Values.agent.otelConfig) . ) }}
+  {{- if $customAgent.prometheus }}
+  {{- if $customAgent.prometheus.config }}
+  prometheus:
+    {{- with $customAgent.prometheus.config }}
+    config:
+      {{- toYaml . | nindent 6 }}
+    {{- end }}
```
Comment (musa-asad) [inline charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-custom-resource.yaml:91]: We need to add this to allow the user to use the agent values.
```suggestion
    {{- end }}
    {{- else if $.Values.agent.prometheus.config }}
  prometheus:
    {{- with $.Values.agent.prometheus.config }}
    config:
      {{- toYaml . | nindent 6 }}
    {{- end }}
```

File: charts/amazon-cloudwatch-observability/values.yaml [hunk 3]
```diff
@@ -1185,8 +1185,14 @@ admissionWebhooks:
...
+agents:
+  - name: cloudwatch-agent
+
 agent:
   name:
+  mode: # represents the mode the cloudwatch-agent will run in (deployment, daemonset or statefulset)
+  replicas: # The total number non-terminated pods targeted by this cloudwatch-agent's deployment or statefulSet.
   image:
     repository: cloudwatch-agent
     tag: 1.300049.1b929
```
Comment (okankoAMZ) [inline charts/amazon-cloudwatch-observability/values.yaml:1198]: need to update this

**Review by okankoAMZ — COMMENTED** [review]
we need to update: operator tag:tag: 1.9.0

File: charts/amazon-cloudwatch-observability/values.yaml [hunk 4]
```diff
@@ -1243,6 +1249,37 @@ agent:
...
+        }
+      }
+    }
+  prometheus:
+    config:
+    targetAllocator:
+      enabled: false
+      image:
+        repository: cloudwatch-agent-target-allocator
+        tag: latest
```
Comment (sky333999) [inline charts/amazon-cloudwatch-observability/values.yaml:1274]: This should be 1.0.0

File: charts/amazon-cloudwatch-observability/values.yaml [hunk 5]
```diff
@@ -1243,6 +1249,37 @@ agent:
         }
       }
     }
+  windowsDefaultConfig:
```
Comment (sky333999) [inline charts/amazon-cloudwatch-observability/values.yaml:1252]: Dont think this property is used anywhere.. we should remove it

Comment (mitali-salvi) [reply charts/amazon-cloudwatch-observability/values.yaml:1252]: yep this should be removed

File: charts/amazon-cloudwatch-observability/values.yaml (hunk 4 above)
Comment (mitali-salvi) [reply charts/amazon-cloudwatch-observability/values.yaml:1274]: +1

File: charts/amazon-cloudwatch-observability/values.yaml (hunk 3 above)
Comment (sky333999) [reply charts/amazon-cloudwatch-observability/values.yaml]: Being done as part of https://github.com/aws-observability/helm-charts/pull/142

**Review by sky333999 — APPROVED** [review]

**Comment by sky333999** [comment]
There's room for improvements with how we handle the overlays for `agents` on top of the default `agent`, but that can be a follow-up.

Did a helm template render of the chart with and without this change and the diffs look as expected - essentially a no-op in the default case.

**Review by mitali-salvi — APPROVED** [review]

**Review by musa-asad — APPROVED** [review]

**Comment by David-Crty** [comment]
I don't understand why this is merged into the main, but no new releases are done. 

How can we know if this available on release 2.6.0 or not ?

**Pull request merged**